In [30]:
import pandas as pd 
order=pd.read_csv(r'C:\Users\Shubham\Project\Ecommerce Profitability Analysis\data\orders.csv')
products=pd.read_csv(r'C:\Users\Shubham\Project\Ecommerce Profitability Analysis\data\products.csv')
marketing=pd.read_csv(r'C:\Users\Shubham\Project\Ecommerce Profitability Analysis\data\marketing_spend.csv')

print(f"Order:{order.shape} having columns {order.columns.to_list()}")
print(f"Products:{products.shape} having columns {products.columns.to_list()}")
print(f"Marketing:{marketing.shape} having columns {marketing.columns.to_list()}")

print(f"Order head:{order.head(3)}")
print(f"Products head:{products.head(3)}")
print(f"Marketing head:{marketing.head(3)}")


Order:(2000, 20) having columns ['order_id', 'customer_id', 'order_date', 'channel', 'payment_method', 'region', 'items_ordered', 'primary_category', 'gross_revenue', 'discount_pct', 'discount_amount', 'shipping_cost', 'product_cost', 'platform_fee', 'transaction_fee', 'returned', 'refund_amount', 'net_revenue', 'total_costs', 'profit']
Products:(207, 9) having columns ['product_id', 'product_name', 'category', 'sub_category', 'unit_cost', 'selling_price', 'shipping_cost_per_unit', 'weight_lbs', 'supplier']
Marketing:(144, 10) having columns ['month', 'platform', 'spend', 'impressions', 'clicks', 'conversions', 'revenue_attributed', 'cpc', 'cpa', 'roas']
Order head:     order_id customer_id  order_date     channel payment_method     region  \
0  ORD-000001      C-0617  2025-06-24  Mobile App      Gift Card  Northeast   
1  ORD-000002      C-0720  2025-08-23     Website    Credit Card    Midwest   
2  ORD-000003      C-0189  2025-07-20     Website    Credit Card  Southeast   

   items_

In [31]:
print(f"Order Nulls:{order.isnull().sum()}")
print(f"products Nulls:{products.isnull().sum()}")
print(f"Marketing Nulls:{marketing.isnull().sum()}")

print(f"Order Duplicates:{order.duplicated().sum()}")
print(f"Prodcuts Duplicates:{products.duplicated().sum()}")
print(f"Marketing Duplicates:{marketing.duplicated().sum()}")

print(f"Order Dtypes:{order.dtypes}")
print(f"Products Dtypes:{products.dtypes}")
print(f"marketing Dtypes:{marketing.dtypes}")

Order Nulls:order_id            0
customer_id         0
order_date          0
channel             0
payment_method      0
region              0
items_ordered       0
primary_category    0
gross_revenue       0
discount_pct        0
discount_amount     0
shipping_cost       0
product_cost        0
platform_fee        0
transaction_fee     0
returned            0
refund_amount       0
net_revenue         0
total_costs         0
profit              0
dtype: int64
products Nulls:product_id                0
product_name              0
category                  0
sub_category              0
unit_cost                 0
selling_price             0
shipping_cost_per_unit    0
weight_lbs                0
supplier                  0
dtype: int64
Marketing Nulls:month                 0
platform              0
spend                 0
impressions           0
clicks                0
conversions           0
revenue_attributed    0
cpc                   0
cpa                   0
roas                  0

In [32]:
# Basic Stats
print("---Order stats---")
print(order[[
    'gross_revenue', 'discount_pct', 'shipping_cost','product_cost', 'platform_fee', 'profit','refund_amount'
]].describe())

print("\n --Marketing Stats--")
print(marketing[['spend', 'revenue_attributed','roas', 'cpc', 'cpa']].describe())

---Order stats---
       gross_revenue  discount_pct  shipping_cost  product_cost  platform_fee  \
count    2000.000000   2000.000000    2000.000000   2000.000000   2000.000000   
mean      138.984565      7.635000      25.552930     55.466975      4.947310   
std       128.558780      7.857331      18.195304     48.014324     11.936644   
min         3.490000      0.000000       3.740000      2.170000      0.000000   
25%        49.350000      0.000000      11.580000     20.880000      0.000000   
50%        99.995000      5.000000      20.750000     40.965000      0.000000   
75%       194.262500     15.000000      35.335000     77.022500      4.027500   
max      1081.140000     25.000000     111.240000    364.930000    162.170000   

            profit  refund_amount  
count  2000.000000    2000.000000  
mean     28.166915      10.291225  
std      69.551171      53.474418  
min    -469.090000       0.000000  
25%      -2.357500       0.000000  
50%      15.185000       0.000000  


In [33]:
# ── Unique values in categorical columns ──
print("Channels      :", order['channel'].unique())
print("Categories    :", order['primary_category'].unique())
print("Returned      :", order['returned'].unique())
print("Regions       :", order['region'].unique())
print("Platforms     :", marketing['platform'].unique())
print("Payment methods:", order['payment_method'].unique())

Channels      : <StringArray>
['Mobile App', 'Website', 'Marketplace', 'Social Commerce']
Length: 4, dtype: str
Categories    : <StringArray>
[    'Electronics',          'Beauty',            'Toys',          'Sports',
        'Clothing', 'Food & Beverage',           'Books',  'Home & Kitchen']
Length: 8, dtype: str
Returned      : <StringArray>
['No', 'Yes']
Length: 2, dtype: str
Regions       : <StringArray>
['Northeast', 'Midwest', 'Southeast', 'West Coast', 'Southwest']
Length: 5, dtype: str
Platforms     : <StringArray>
[     'Google Ads',    'Facebook Ads',   'Instagram Ads',      'TikTok Ads',
 'Email Marketing',      'Influencer']
Length: 6, dtype: str
Payment methods: <StringArray>
['Gift Card', 'Credit Card', 'Debit Card', 'Buy Now Pay Later', 'PayPal']
Length: 5, dtype: str


In [34]:
order["order_date"]=pd.to_datetime(order['order_date'])

print("Order date range::")
print(f"From:{order['order_date'].min()}")
print(f"To :{order['order_date'].max()}")

print("\n Marketing months")
print(marketing['month'].unique())

Order date range::
From:2024-01-28 00:00:00
To :2026-01-11 00:00:00

 Marketing months
<StringArray>
['2024-01', '2024-02', '2024-03', '2024-04', '2024-05', '2024-06', '2024-07',
 '2024-08', '2024-09', '2024-10', '2024-11', '2024-12', '2025-01', '2025-02',
 '2025-03', '2025-04', '2025-05', '2025-06', '2025-07', '2025-08', '2025-09',
 '2025-10', '2025-11', '2025-12']
Length: 24, dtype: str


In [35]:
# Verifing gross_revenue-total_costs=profit

order['profit_check']=(order['gross_revenue']-order['total_costs']).round(2)

mismatch=order[order['profit_check']!=order['profit'].round(2)]
print(f"Profit calculation mismatches:{len(mismatch)}")

returned_with_revenue=order[(order['returned']=='Yes')&(order['net_revenue']>0)]


print(f"Returned orders with positive net revenue:{len(returned_with_revenue)}")

Profit calculation mismatches:1257
Returned orders with positive net revenue:0


In [36]:
# Look  mismatch rows closely
mismatch = order[order['profit_check'] != order['profit'].round(2)].copy()

print(f"Total mismatches: {len(mismatch)}")
print(f"\nSample mismatches:")
print(mismatch[['order_id', 'gross_revenue', 'total_costs', 
 'profit', 'profit_check', 'returned', 'refund_amount', 'discount_amount']].head(10).to_string())


print("Mismatches by returned status:")
print(mismatch['returned'].value_counts())


print("\nNon-returned mismatches:")
non_returned_mismatch = mismatch[mismatch['returned'] == 'No']
print(non_returned_mismatch[['order_id', 'gross_revenue', 'total_costs', 'profit', 'profit_check']].head(5).to_string())



# Check the difference amount
mismatch['difference'] = (mismatch['profit'] - mismatch['profit_check']).round(2)

print("Difference stats:")
print(mismatch['difference'].describe())
print("\nUnique difference values (first 20):")
print(mismatch['difference'].value_counts().head(20))


# Checking if refund amount explains the difference
mismatch['refund_explains'] = (
    mismatch['profit_check'] - mismatch['refund_amount']
).round(2) == mismatch['profit'].round(2)

print("Does refund_amount explain the difference?")
print(mismatch['refund_explains'].value_counts())


Total mismatches: 1257

Sample mismatches:
      order_id  gross_revenue  total_costs  profit  profit_check returned  refund_amount  discount_amount
3   ORD-000004          30.63        48.39  -19.29        -17.76       No           0.00             1.53
4   ORD-000005          11.99        19.09  -10.10         -7.10       No           0.00             3.00
5   ORD-000006         126.99       122.73  -14.79          4.26       No           0.00            19.05
7   ORD-000008          71.03        72.44  -72.44         -1.41      Yes          63.93             7.10
8   ORD-000009          23.98        35.25  -14.87        -11.27       No           0.00             3.60
9   ORD-000010         345.27       134.32  124.63        210.95       No           0.00            86.32
10  ORD-000011          26.50        18.37    6.81          8.13       No           0.00             1.32
12  ORD-000013         136.77       102.46   20.63         34.31       No           0.00            13.68
14 

In [37]:
# Test hypothesis: profit = gross_revenue - total_costs - discount_amount
order['profit_test'] = (
    order['gross_revenue'] - order['total_costs'] - order['discount_amount']
).round(2)

match = (order['profit_test'] == order['profit'].round(2)).sum()
print(f"Matches with discount subtracted: {match} / {len(order)}")

# Test for returned orders specifically
order['profit_test2'] = (
    order['gross_revenue'] - order['total_costs'] - 
    order['discount_amount'] - order['refund_amount']
).round(2)

match2 = (order['profit_test2'] == order['profit'].round(2)).sum()
print(f"Matches with discount + refund subtracted: {match2} / {len(order)}")

Matches with discount subtracted: 1856 / 2000
Matches with discount + refund subtracted: 2000 / 2000


In [38]:

order['profit_check'] = (
    order['gross_revenue'] - order['total_costs'] - 
    order['discount_amount'] - order['refund_amount']
).round(2)

# V should be 2000/2000 now
match = (order['profit_check'] == order['profit'].round(2)).sum()
print(f"Profit verification: {match} / {len(order)} match")

# Use existing profit column — it's correct
print("\nSample verification:")
print(order[['order_id', 'gross_revenue', 'discount_amount', 
              'refund_amount', 'total_costs', 
              'profit']].head(5).to_string())

Profit verification: 2000 / 2000 match

Sample verification:
     order_id  gross_revenue  discount_amount  refund_amount  total_costs  profit
0  ORD-000001         261.39             0.00            0.0       122.46  138.93
1  ORD-000002          24.02             0.00            0.0        15.73    8.29
2  ORD-000003          52.60             0.00            0.0        47.84    4.76
3  ORD-000004          30.63             1.53            0.0        48.39  -19.29
4  ORD-000005          11.99             3.00            0.0        19.09  -10.10


In [39]:
# ── Negative profit investigation ──
neg_profit = order[order['profit'] < 0]
print(f"\n=== NEGATIVE PROFIT ORDERS ===")
print(f"Count: {len(neg_profit)} / {len(order)}")
print(f"Total loss from negative profit orders: ${neg_profit['profit'].sum():,.2f}")
print("\nNegative profit by category:")
print(neg_profit['primary_category'].value_counts())


=== NEGATIVE PROFIT ORDERS ===
Count: 563 / 2000
Total loss from negative profit orders: $-17,911.97

Negative profit by category:
primary_category
Books              109
Sports              85
Clothing            83
Food & Beverage     75
Beauty              72
Electronics         57
Toys                51
Home & Kitchen      31
Name: count, dtype: int64


In [40]:
# ── Nulls ──
print("=== NULLS ===")
print("Order nulls:")
print(order.isnull().sum()[order.isnull().sum() > 0])
print("\nProducts nulls:")
print(products.isnull().sum()[products.isnull().sum() > 0])
print("\nMarketing nulls:")
print(marketing.isnull().sum()[marketing.isnull().sum() > 0])
# ── Duplicates ──
print("=== DUPLICATES ===")
print(f"Order      : {order.duplicated().sum()}")
print(f"Products   : {products.duplicated().sum()}")
print(f"Marketing  : {marketing.duplicated().sum()}")
# ── Categorical checks ──
print("=== UNIQUE VALUES ===")
print("Channels        :", order['channel'].unique())
print("Categories      :", order['primary_category'].unique())
print("Returned        :", order['returned'].unique())
print("Regions         :", order['region'].unique())
print("Platforms       :", marketing['platform'].unique())
print("Payment methods :", order['payment_method'].unique())
# ── Date check ──
print("=== DATE RANGE ===")
print(f"Orders from : {order['order_date'].min().date()}")
print(f"Orders to   : {order['order_date'].max().date()}")
print(f"\nMarketing covers: {marketing['month'].min()} → {marketing['month'].max()}")

future = order[order['order_date'] > '2025-12-31']
print(f"\nOrders after Dec 2025: {len(future)}")
if len(future) > 0:
    print(future['order_date'].value_counts().head())
# ── Key stats ──
print("=== ORDER KEY STATS ===")
print(order[['gross_revenue', 'profit', 'discount_pct',
             'shipping_cost', 'platform_fee',
             'refund_amount']].describe().round(2))

=== NULLS ===
Order nulls:
Series([], dtype: int64)

Products nulls:
Series([], dtype: int64)

Marketing nulls:
Series([], dtype: int64)
=== DUPLICATES ===
Order      : 0
Products   : 0
Marketing  : 0
=== UNIQUE VALUES ===
Channels        : <StringArray>
['Mobile App', 'Website', 'Marketplace', 'Social Commerce']
Length: 4, dtype: str
Categories      : <StringArray>
[    'Electronics',          'Beauty',            'Toys',          'Sports',
        'Clothing', 'Food & Beverage',           'Books',  'Home & Kitchen']
Length: 8, dtype: str
Returned        : <StringArray>
['No', 'Yes']
Length: 2, dtype: str
Regions         : <StringArray>
['Northeast', 'Midwest', 'Southeast', 'West Coast', 'Southwest']
Length: 5, dtype: str
Platforms       : <StringArray>
[     'Google Ads',    'Facebook Ads',   'Instagram Ads',      'TikTok Ads',
 'Email Marketing',      'Influencer']
Length: 6, dtype: str
Payment methods : <StringArray>
['Gift Card', 'Credit Card', 'Debit Card', 'Buy Now Pay Later', 'P

In [41]:
# ── Fix date type ──
order['order_date'] = pd.to_datetime(order['order_date'])

# ── Fix marketing month to datetime for time series later ──
marketing['month_dt'] = pd.to_datetime(marketing['month'])

# ── Add profit margin % column ──
order['profit_margin_%'] = (
    order['profit'] / order['gross_revenue'] * 100
).round(2)

# ── Add binary returned column for easier aggregation ──
order['returned_binary'] = (order['returned'] == 'Yes').astype(int)

# ── Add month column for time series joining with marketing ──
order['month'] = order['order_date'].dt.to_period('M').astype(str)

# ── Filter out 2026 orders — marketing data only covers 2024-2025 ──
order_clean = order[order['order_date'] <= '2025-12-31'].copy()

print(f"Original order count  : {len(order)}")
print(f"Clean order count     : {len(order_clean)}")
print(f"Removed (2026 dates)  : {len(order) - len(order_clean)}")
print(f"\nNew columns added:")
print(order_clean[['profit_margin_%', 'returned_binary', 
                    'month']].head(3))

Original order count  : 2000
Clean order count     : 1997
Removed (2026 dates)  : 3

New columns added:
   profit_margin_%  returned_binary    month
0            53.15                0  2025-06
1            34.51                0  2025-08
2             9.05                0  2025-07


In [ ]:
order_clean.to_csv('../data/order_clean.csv', index=False)
products.to_csv('../data/products_clean.csv', index=False)
marketing.to_csv('../data/marketing_clean.csv', index=False)

print(" All clean files exported")
print(f"order_clean    : {len(order_clean)} rows")
print(f"products_clean : {len(products)} rows")
print(f"marketing_clean: {len(marketing)} rows")

✅ All clean files exported
order_clean    : 1997 rows
products_clean : 207 rows
marketing_clean: 144 rows


## Data Quality Summary — BrightCart Analysis

| Check | Order | Products | Marketing |
|---|---|---|---|
| Null values | 0 ✅ | 0 ✅ | 0 ✅ |
| Duplicates | 0 ✅ | 0 ✅ | 0 ✅ |
| Type fixes | order_date → datetime ✅ | — | month → datetime ✅ |
| Outliers | None requiring removal ✅ | — | — |
| Future dates | 2026 orders removed | — | — |

## Key Pre-Analysis Findings

- **563 / 2000 orders (28.15%) have negative profit** — primary driver
  of shrinking margins
- **Profit formula confirmed:** 
  gross_revenue - total_costs - discount_amount - refund_amount
- **4 channels:** Website, Mobile App, Marketplace, Social Commerce
- **8 categories:** Electronics, Beauty, Toys, Sports, Clothing,
  Food & Beverage, Books, Home & Kitchen
- **Marketing covers:** Jan 2024 – Dec 2025 (24 months, 12 platforms)
- **Analysis period:** Jan 2024 – Dec 2025

## Files Exported
- order_clean.csv
- products_clean.csv  
- marketing_clean.csv